readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/kafka/OffsetResetter.java

```diff
@@ -66,32 +66,39 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
                 final Map<Integer, Long> startingOffsets = writer.getStartingOffsets(partitionsId);
 
                 startingOffsets.forEach((part, offset) ->
-                        offsetsPerPartition.computeIfAbsent(part, ignored -> new ArrayList<>()).add(offset));
+                        offsetsPerPartition.computeIfAbsent(part, ignored -> new HashSet<>()).add(offset));
             } catch (IOException e) {
                 LOGGER.warn("Couldn't get offset for partitions {}, will resume from earliest",
                         partitionsId.stream().map(String::valueOf).collect(Collectors.joining(", ")), e);
                 partitions.forEach(part ->
-                        offsetsPerPartition.computeIfAbsent(part.partition(), ignored -> new ArrayList<>())
+                        offsetsPerPartition.computeIfAbsent(part.partition(), ignored -> new HashSet<>())
                         .add(UNKNOWN_OFFSET));
                 // Don't break here as we need all exceptional writers to cache "unknown offset" for future queries
             }
         }
 
+        //Compute min offset per partition, taking into account the possibility
+        //of a new event that would reset all partitions to UNKNOWN_OFFSET
+        //in this case keep the lower offset above
+
+        boolean possibleNewEvent = offsetsPerPartition.values().stream().allMatch(offsets -> offsets.contains(UNKNOWN_OFFSET));
+
+        Map<Integer, Long> minOffsetPerPartition = offsetsPerPartition
+            .entrySet()
+            .stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, offsets -> offsets
+                .getValue()
+                .stream()
+                .filter(offset -> offset != UNKNOWN_OFFSET || !possibleNewEvent)
```

Comment:
What about filtering unknown offsets, then only resetting if the resulting list is empty? This would seem more legible to me.

Reply:
agreed that         boolean possibleNewEvent = offsetsPerPartition.values().stream().allMatch(offsets -> offsets.contains(UNKNOWN_OFFSET));
 is not relevant since we only see partitions for this reader.